File: server/src/state.js
class Slot {
  constructor({ id, row, col, zone, status = 0, type = "standard", price = 2.5 }) {
    this.id = id;
    this.row = row;
    this.col = col;
    this.zone = zone;
    this.status = status; // 0 vacant, 1 occupied, 2 unavailable
    this.type = type;
    this.price = price;
    this.updatedAt = new Date();
  }

  toJSON() {
    return {
      id: this.id,
      row: this.row,
      col: this.col,
      zone: this.zone,
      status: this.status,
      type: this.type,
      price: Number(this.price.toFixed(2)),
      updated_at: this.updatedAt.toISOString(),
    };
  }

  toRecord() {
    return {
      id: this.id,
      row: this.row,
      col: this.col,
      zone: this.zone,
      status: this.status,
      type: this.type,
      price: this.price,
      updatedAt: this.updatedAt,
    };
  }
}

export class GridState {
  constructor(rows = 8, cols = 10) {
    this.rows = rows;
    this.cols = cols;
    this.lotId = "central-lot";
    this.slots = new Map();
  }

  seed() {
    const zones = ["A", "B"];
    for (let r = 0; r < this.rows; r += 1) {
      for (let c = 0; c < this.cols; c += 1) {
        const slotId = `S-${r}-${c}`;
        const zone = zones[r % zones.length];
        const type = Math.random() < 0.15 ? "ev" : Math.random() < 0.25 ? "accessible" : "standard";
        this.slots.set(
          slotId,
          new Slot({
            id: slotId,
            row: r,
            col: c,
            zone,
            status: 0,
            type,
            price: 2.5,
          })
        );
      }
    }
  }

  async loadFromModel(SlotModel) {
    const docs = await SlotModel.find({});
    if (!docs.length) {
      this.seed();
      await SlotModel.insertMany(Array.from(this.slots.values()).map((s) => s.toRecord()));
      return;
    }
    this.slots.clear();
    docs.forEach((doc) => {
      this.slots.set(
        doc.id,
        new Slot({
          id: doc.id,
          row: doc.row,
          col: doc.col,
          zone: doc.zone,
          status: doc.status,
          type: doc.type,
          price: doc.price,
        })
      );
    });
  }

  snapshot() {
    return {
      lot_id: this.lotId,
      rows: this.rows,
      cols: this.cols,
      slots: Array.from(this.slots.values()).map((s) => s.toJSON()),
    };
  }

  occupancy(zone) {
    const slots = Array.from(this.slots.values()).filter((s) => !zone || s.zone === zone);
    if (!slots.length) return 0;
    const occupied = slots.filter((s) => s.status === 1).length;
    return occupied / slots.length;
  }

  setStatus(slotId, status, price) {
    if (!this.slots.has(slotId)) throw new Error("not-found");
    const slot = this.slots.get(slotId);
    slot.status = status;
    if (price !== undefined) slot.price = price;
    slot.updatedAt = new Date();
    return slot;
  }

  async persistSlot(SlotModel, slot) {
    await SlotModel.findOneAndUpdate(
      { id: slot.id },
      {
        row: slot.row,
        col: slot.col,
        zone: slot.zone,
        status: slot.status,
        type: slot.type,
        price: slot.price,
        updatedAt: slot.updatedAt,
      },
      { upsert: true, new: true }
    );
  }

  bulkUpdate(slotIds, status, price) {
    const updated = [];
    slotIds.forEach((id) => {
      if (!this.slots.has(id)) return;
      const slot = this.slots.get(id);
      if (status !== undefined) slot.status = status;
      if (price !== undefined) slot.price = price;
      slot.updatedAt = new Date();
      updated.push(slot);
    });
    return updated;
  }

  async persistBulk(SlotModel, slots) {
    const ops = slots.map((slot) => ({
      updateOne: {
        filter: { id: slot.id },
        update: {
          row: slot.row,
          col: slot.col,
          zone: slot.zone,
          status: slot.status,
          type: slot.type,
          price: slot.price,
          updatedAt: slot.updatedAt,
        },
        upsert: true,
      },
    }));
    if (ops.length) {
      await SlotModel.bulkWrite(ops);
    }
  }
}

export const grid = new GridState();

File: backend/app/state.py
from __future__ import annotations

import asyncio
import random
from dataclasses import dataclass, field
from datetime import datetime
from typing import Dict, List, Optional


@dataclass
class Slot:
    id: str
    row: int
    col: int
    zone: str
    status: int  # 0 = vacant, 1 = occupied, 2 = unavailable
    type: str = "standard"  # standard | ev | accessible
    price: float = 2.5
    updated_at: datetime = field(default_factory=datetime.utcnow)

    def to_dict(self) -> Dict:
        return {
            "id": self.id,
            "row": self.row,
            "col": self.col,
            "zone": self.zone,
            "status": self.status,
            "type": self.type,
            "price": round(self.price, 2),
            "updated_at": self.updated_at.isoformat() + "Z",
        }


class GridState:
    """In-memory digital twin for a single lot."""

    def __init__(self, rows: int = 8, cols: int = 10):
        self.rows = rows
        self.cols = cols
        self.slots: Dict[str, Slot] = {}
        self.lot_id = "central-lot"
        self._lock = asyncio.Lock()
        self._seed()

    def _seed(self):
        zones = ["A", "B"]
        for r in range(self.rows):
            for c in range(self.cols):
                slot_id = f"S-{r}-{c}"
                zone = zones[r % len(zones)]
                slot_type = random.choice(["standard"] * 6 + ["ev", "accessible"])
                self.slots[slot_id] = Slot(
                    id=slot_id,
                    row=r,
                    col=c,
                    zone=zone,
                    status=0,
                    type=slot_type,
                    price=2.5,
                )

    async def set_status(self, slot_id: str, status: int, price: Optional[float] = None):
        async with self._lock:
            if slot_id not in self.slots:
                raise KeyError(slot_id)
            slot = self.slots[slot_id]
            slot.status = status
            if price is not None:
                slot.price = price
            slot.updated_at = datetime.utcnow()
            return slot

    async def bulk_update(
        self, slot_ids: List[str], status: Optional[int] = None, price: Optional[float] = None
    ) -> List[Slot]:
        updated = []
        async with self._lock:
            for slot_id in slot_ids:
                if slot_id not in self.slots:
                    continue
                slot = self.slots[slot_id]
                if status is not None:
                    slot.status = status
                if price is not None:
                    slot.price = price
                slot.updated_at = datetime.utcnow()
                updated.append(slot)
        return updated

    def snapshot(self) -> Dict:
        return {
            "lot_id": self.lot_id,
            "rows": self.rows,
            "cols": self.cols,
            "slots": [s.to_dict() for s in self.slots.values()],
        }

    def occupancy(self, zone: Optional[str] = None) -> float:
        slots = [s for s in self.slots.values() if zone is None or s.zone == zone]
        if not slots:
            return 0.0
        occupied = sum(1 for s in slots if s.status == 1)
        return occupied / len(slots)


grid = GridState()

